api/envoy/type/tracing/v2/custom_tag.proto

```diff
@@ -0,0 +1,67 @@
+syntax = "proto3";
+
+package envoy.type.tracing.v2;
+
+option java_outer_classname = "CustomTagProto";
+option java_multiple_files = true;
+option java_package = "io.envoyproxy.envoy.type.tracing.v2";
+
+import "envoy/type/matcher/metadata.proto";
+
+import "validate/validate.proto";
+
+// [#protodoc-title: Custom Tag]
+
+// Describes custom tags for the active span.
+message CustomTag {
+  // Literal type custom tag with static value for the tag value.
+  message Literal {
+    // Static literal value to populate the tag value.
+    string value = 1 [(validate.rules).string = {min_bytes: 1}];
+  }
+
+  // Environment type custom tag with environment name and default value.
+  message Environment {
+    // Environment variable name to obtain the value to populate the tag value.
+    string name = 1 [(validate.rules).string = {min_bytes: 1}];
+
+    // When the environment variable is not found,
+    // the tag value will be populated with this default value if specified,
+    // otherwise no tag will be populated.
+    string default_value = 2;
+  }
+
+  // Header type custom tag with header name and default value.
+  message Header {
+    // Request header name to obtain the value to populate the tag value.
+    string name = 1 [(validate.rules).string = {min_bytes: 1}];
+
+    // When the request header does not exist,
+    // the tag value will be populated with this default value if specified,
+    // otherwise no tag will be populated.
+    string default_value = 2;
+  }
+
+  // Used to populate the tag name.
+  string tag = 1 [(validate.rules).string = {min_bytes: 1}];
+
+  // Used to specify what kind of custom tag.
+  oneof type {
+    option (validate.required) = true;
+
+    // A literal custom tag.
+    Literal literal = 2;
+
+    // An environment custom tag.
+    Environment environment = 3;
+
+    // A request header custom tag.
+    Header request_header = 4;
+
+    // A metadata custom tag associated with the request.
```

Comment:
Sorry for being dense, but I find metadata to be very confusing. What exactly is the difference between request metadata and metadata? Can you link to and/or describe the relevant metadata that this matches? Also, don't you actually need a combination of matcher as well as default value for each option?

cc @lizan @kyessenov can you take a look at this and see what you think?

Reply:
I also think that almost each type has a default value that is redundant. But the default value is at least unreasonable for static literal types. It's ok to move it out to the top level though.

Reply:
I still prefer having the `Metadata` message with the matcher and default value defined in there. Just in case other custom tag sources may be defined in the future, better to have the criteria for each encapsulated in the sub types?

Reply:
@objectiser Yup, `PathSegment` in `MetadataMatcher` may be a good reference though.

Reply:
> I still prefer having the Metadata message with the matcher and default value defined in there. Just in case other custom tag sources may be defined in the future, better to have the criteria for each encapsulated in the sub types?

+1 that's what I had in mind also.